crates/net/downloaders/src/bodies/concurrent.rs

```diff
@@ -48,430 +218,376 @@ where
+                        self.in_progress_queue.push_new_request(
+                            Arc::clone(&self.client),
+                            Arc::clone(&self.consensus),
+                            self.query_headers(request_range.start, request_range.count())?,
+                        );
+                        // Clear the current request range
+                        request_range = Range::default();
+                    }
+                }
+            }
+        }
+
+        // Put queued bodies into the buffer
+        if !queued_bodies.is_empty() {
+            self.buffered_responses.push(OrderedBodiesResponse(queued_bodies));
+        }
+
+        self.download_range = range;
+        self.latest_queued_block_number = None;
+        tracing::trace!(target: "downloaders::bodies", range = ?self.download_range, "New header range set");
+        Ok(())
     }
 }
 
-impl<Client, Consensus> ConcurrentDownloader<Client, Consensus>
+impl<B, C, DB> Stream for ConcurrentDownloader<B, C, DB>
 where
-    Client: BodiesClient,
-    Consensus: ConsensusTrait,
+    B: BodiesClient + 'static,
+    C: ConsensusTrait + 'static,
+    DB: Database,
 {
-    /// Create a new concurrent downloader instance.
-    pub fn new(client: Arc<Client>, consensus: Arc<Consensus>) -> Self {
-        Self { client, consensus, retries: 3, batch_size: 100, concurrency: 5 }
+    type Item = Vec<BlockResponse>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        if this.is_terminated() {
+            return Poll::Ready(None)
+        }
+
+        // Submit new requests and poll any in progress
+        loop {
+            // Poll requests
+            while let Poll::Ready(Some(response)) = this.in_progress_queue.poll_next_unpin(cx) {
+                let response = OrderedBodiesResponse(response);
+                this.buffered_responses.push(response);
+            }
+
+            // Loop exit condition
+            let mut new_request_submitted = false;
+            // Submit new requests
+            let concurrent_requests_limit = this.concurrent_request_limit();
+            'inner: while this.in_progress_queue.len() < concurrent_requests_limit &&
+                this.buffered_responses.len() < this.max_buffered_responses
+            {
+                match this.next_headers_request() {
+                    Ok(Some(request)) => {
+                        this.in_progress_queue.push_new_request(
+                            Arc::clone(&this.client),
+                            Arc::clone(&this.consensus),
+                            request,
+                        );
+                        new_request_submitted = true;
+                    }
+                    Ok(None) => break 'inner,
+                    Err(error) => {
+                        tracing::error!(target: "downloaders::bodies", ?error, "Failed to form next request");
+                    }
```

Comment:
Are we ok with not failing in any way here?

Reply:
up for a discussion. trying to think if it can be a recoverable error...

Reply:
this is only possible if there's an issue with the DB, right?

this is critical, so perhaps this should be propagated.